File: ios/PocDedicatedScanner/BluetoothAccessoryPicker.m
#import <React/RCTBridgeModule.h>
#import <ExternalAccessory/ExternalAccessory.h>

@interface BluetoothAccessoryPicker : NSObject <RCTBridgeModule>
@end

@implementation BluetoothAccessoryPicker

RCT_EXPORT_MODULE()

RCT_EXPORT_METHOD(showPicker:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  dispatch_async(dispatch_get_main_queue(), ^{
    [[EAAccessoryManager sharedAccessoryManager]
      showBluetoothAccessoryPickerWithNameFilter:nil
      completion:^(NSError *error) {
        if (!error || error.code == EABluetoothAccessoryPickerAlreadyConnected) {
          resolve(nil);
        } else if (error.code == EABluetoothAccessoryPickerResultCancelled) {
          reject(@"USER_CANCELED", error.localizedDescription, error);
        } else if (error.code == EABluetoothAccessoryPickerResultNotFound) {
          reject(@"NO_ACCESSORIES_FOUND", error.localizedDescription, error);
        } else {
          reject(@"NOT_SUPPORTED", error.localizedDescription, error);
        }
      }];
  });
}

RCT_EXPORT_METHOD(getConnectedAccessories:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSArray<EAAccessory *> *accessories = [[EAAccessoryManager sharedAccessoryManager] connectedAccessories];
  NSMutableArray *result = [NSMutableArray array];
  for (EAAccessory *acc in accessories) {
    [result addObject:@{
      @"name": acc.name ?: @"",
      @"manufacturer": acc.manufacturer ?: @"",
      @"modelNumber": acc.modelNumber ?: @"",
      @"serialNumber": acc.serialNumber ?: @"",
      @"firmwareRevision": acc.firmwareRevision ?: @"",
      @"hardwareRevision": acc.hardwareRevision ?: @"",
      @"connectionID": @(acc.connectionID),
      @"protocolStrings": acc.protocolStrings ?: @[],
    }];
  }
  resolve(result);
}

+ (BOOL)requiresMainQueueSetup { return NO; }

@end

File: ios/PocDedicatedScanner/BluetoothAccessoryPicker.swift
import Foundation
import ExternalAccessory

@objc(BluetoothAccessoryPicker)
class BluetoothAccessoryPicker: NSObject {

  @objc func showPicker(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      EAAccessoryManager.shared().showBluetoothAccessoryPicker(withNameFilter: nil) { error in
        guard let error = error as NSError? else {
          resolve(nil)
          return
        }
        let code: String
        switch EABluetoothAccessoryPickerError.Code(rawValue: error.code) {
        case .alreadyConnected:
          // Treat as success — already paired
          resolve(nil)
          return
        case .resultNotFound:
          code = "NO_ACCESSORIES_FOUND"
        case .resultCancelled:
          code = "USER_CANCELED"
        case .resultFailed:
          code = "NOT_SUPPORTED"
        default:
          code = "UNKNOWN"
        }
        reject(code, error.localizedDescription, error)
      }
    }
  }

  @objc static func requiresMainQueueSetup() -> Bool { return false }
}
